packages/ovm/src/contracts/ExecutionManager.sol

```diff
@@ -1064,4 +1022,65 @@ contract ExecutionManager is FullStateManager {
+     * @param _timestamp The timestamp which should be used for this context.
+     * @param _queueOrigin The queue which this context's transaction was sent from.
+     * @param _ovmTxOrigin The tx.origin for the currently executing transaction. It will be ZERO_ADDRESS if it's not an EOA call.
+     */
+    function initializeContext(uint _timestamp, uint _queueOrigin, address _ovmTxOrigin, address _l1MsgSender) internal {
+        // First zero out the context for good measure (Note ZERO_ADDRESS is reserved for the genesis contract & initial msgSender)
+        restoreContractContext(ZERO_ADDRESS, ZERO_ADDRESS);
+        // And finally set the timestamp, queue origin, tx origin, and l1MessageSender
+        executionContext.timestamp = _timestamp;
+        executionContext.queueOrigin = _queueOrigin;
+        executionContext.ovmTxOrigin = _ovmTxOrigin;
+        executionContext.l1MessageSender = _l1MsgSender;
+    }
+
+    /**
+     * @notice Change the active contract to be something new. This is used when a new contract is called.
+     * @param _newActiveContract The new active contract
+     * @return The old msgSender and activeContract. This will be used when we restore the old active contract.
+     */
+    function switchActiveContract(address _newActiveContract) internal returns(address _oldMsgSender, address _oldActiveContract) {
+        // Store references to the old context
+        _oldActiveContract = executionContext.ovmActiveContract;
+        _oldMsgSender = executionContext.ovmMsgSender;
+        // Set our new context
+        executionContext.ovmActiveContract = _newActiveContract;
+        executionContext.ovmMsgSender = _oldActiveContract;
+        // Emit an event so we can track the active contract. This is used in order to parse transaction receipts in the fullnode
+        emit ActiveContract(_newActiveContract);
+        // Return old context so we can later revert to it
+        return (_oldMsgSender, _oldActiveContract);
+    }
+
+    /**
+     * @notice Restore the contract context to some old values.
+     * @param _msgSender The msgSender to be restored.
+     * @param _activeContract The activeContract to be restored.
+     */
+    function restoreContractContext(address _msgSender, address _activeContract) internal {
+        // Revert back to the old context
+        executionContext.ovmActiveContract = _activeContract;
+        executionContext.ovmMsgSender = _msgSender;
+    }
+
+    /**
+     * @notice Getter for the execution context's L1MessageSender. Used by the L1MessageSender precompile.
+     * @return The L1MessageSender in our current execution context.
+     */
+    function getL1MessageSender() public returns(address) {
+        require(executionContext.l1MessageSender != ZERO_ADDRESS, "L1MessageSender not set!");
+        require(executionContext.ovmMsgSender == ZERO_ADDRESS, "L1MessageSender only accessible in entrypoint contract!");
```

Comment:
If this is only used by the L1MessageSender precompile, can we be more specific about these `require(...)`s to enforce that?

Reply:
Ah good idea! Will add a require.